File: code/841.keys-and-rooms.php
<?php
/*
 * @lc app=leetcode id=841 lang=php
 *
 * [841] Keys and Rooms
 */

// @lc code=start
class Solution {
    private $visited = [];

    /**
     * @param Integer[][] $rooms
     * @return Boolean
     */
    function canVisitAllRooms($rooms) {
        $this->dfs(0, $rooms);
        return array_sum($this->visited) == count($rooms);
    }

    private function dfs($index, $rooms) {
        $this->visited[$index] = 1;
        foreach ($rooms[$index] as $key) {
            if (isset($this->visited[$key])) continue;
            $this->dfs($key, $rooms);
        }
    }

}
// @lc code=end

File: code/841.keys-and-rooms.go
package code

import "slices"

/*
 * @lc app=leetcode id=841 lang=golang
 *
 * [841] Keys and Rooms
 */

// @lc code=start
func canVisitAllRooms(rooms [][]int) bool {
	visited := make([]bool, len(rooms))
	queue := []int{}

	visited[0] = true
	for _, v := range rooms[0] {
		queue = append(queue, v)
	}

	for len(queue) > 0 {
		key := queue[0]
		queue = queue[1:]

		if visited[key] {
			continue
		}

		visited[key] = true
		for _, v := range rooms[key] {
			queue = append(queue, v)
		}
	}

	return !slices.Contains(visited, false)
}

// @lc code=end
